src/commands/mergecellscommand.js

```diff
@@ -0,0 +1,246 @@
+// @param {module:engine/model/writer~Writer} writer
+function removeRowIfEmpty( row, writer ) {
+	if ( row.childCount ) {
+		return;
+	}
+
+	const table = row.parent;
+	const removedRowIndex = table.getChildIndex( row );
+
+	for ( const { cell, row, rowspan } of new TableWalker( table, { endRow: removedRowIndex } ) ) {
+		const overlapsRemovedRow = row + rowspan - 1 >= removedRowIndex;
+
+		if ( overlapsRemovedRow ) {
+			updateNumericAttribute( 'rowspan', rowspan - 1, cell, writer );
+		}
+	}
+
+	writer.remove( row );
+}
+
+// Merges two table cells - will ensure that after merging cells with empty paragraph the result table cell will only have one paragraph.
+// If one of the merged table cell is empty the merged table cell will have contents of the non-empty table cell.
+// If both are empty the merged table cell will have only one empty paragraph.
+//
+// @param {module:engine/model/element~Element} cellBeingMerged
+// @param {module:engine/model/element~Element} targetCell
+// @param {module:engine/model/writer~Writer} writer
+function mergeTableCells( cellBeingMerged, targetCell, writer ) {
+	if ( !isEmpty( cellBeingMerged ) ) {
+		if ( isEmpty( targetCell ) ) {
+			writer.remove( writer.createRangeIn( targetCell ) );
+		}
+
+		writer.move( writer.createRangeIn( cellBeingMerged ), writer.createPositionAt( targetCell, 'end' ) );
+	}
+
+	// Remove merged table cell.
+	writer.remove( cellBeingMerged );
+}
+
+// Checks if passed table cell contains empty paragraph.
+//
+// @param {module:engine/model/element~Element} tableCell
+// @returns {Boolean}
+function isEmpty( tableCell ) {
+	return tableCell.childCount == 1 && tableCell.getChild( 0 ).is( 'paragraph' ) && tableCell.getChild( 0 ).isEmpty;
+}
+
+// Check if selection contains mergeable cells.
+//
+// In a table below:
+//
+//   +---+---+---+---+
+//   | a | b | c | d |
+//   +---+---+---+   +
+//   | e     | f |   |
+//   +       +---+---+
+//   |       | g | h |
+//   +---+---+---+---+
+//
+// Valid selections are those which creates a solid rectangle (without gaps), such as:
+//   - a, b (two horizontal cells)
+//   - c, f (two vertical cells)
+//   - a, b, e (cell "e" spans over four cells)
+//   - c, d, f (cell d spans over cell in row below)
+//
+// While invalid selection would be:
+//   - a, c (cell "b" not selected creates a gap)
+//   - f, g, h (cell "d" spans over a cell from row of "f" cell - thus creates a gap)
+//
+// @param {module:engine/model/selection~Selection} selection
+// @param {module:table/tableUtils~TableUtils} tableUtils
+// @returns {boolean}
+function canMergeCells( selection, tableUtils ) {
+	// Collapsed selection or selection only one range can't contain mergeable table cells.
+	if ( selection.isCollapsed || selection.rangeCount < 2 ) {
+		return false;
+	}
+
+	const selectedTableCells = getSelectionAffectedTableCells( selection );
+
+	if ( !areCellInTheSameTableSection( selectedTableCells ) ) {
+		return false;
+	}
```

Comment:
Wouldn't simple 
```js
const selectedTableCells = getSelectedTableCells( ... );

if ( selectedTableCells.length < 2 ) {
    return;
} 
```
be enough?

Reply:
Would be enough but checking selection without calculating stuff is IMO a bit faster. The `getSelectedTableCells()` will do all sort of things and here I didn't wanted to calculate all those things.

Probably.

This code is old and I've done some refactoring here already. So WDYT about leaving this as is?

Reply:
1. `getSelectionAffectedTableCells()` executes `getSelectedTableCells()` anyway (see the code). 
2. But `getSelectionAffectedTableCells()` also returns cells that have  a selection **inside** of them (see `getTableCellsContainingSelection()`), which, if not for the initial condition (`selection.isCollapsed || selection.rangeCount < 2)`, would yield invalid cells from the perspective of multi-cell merge. 
3. So considering ☝️ , there's no point in using `getSelectionAffectedTableCells`. `getSelectedTableCells` is enough.

--- 

1. As for the initial condition: I understand that it spares us some unnecessary execution of `getSelectedTableCells()`.
2. But the condition is weak. For instance, you can have a selection with `selection.rangeCount === 3` in FF, each range spanning over some text in a cell, and that condition will pass. 
   2.1. It will also pass if anyone create such selection from code.
   2.2. (referring to the previous section of this comment): in conjunction with `getSelectionAffectedTableCells()` (instead of `getSelectedTableCells()`) this will cause bugs
4. I think it would be better if this condition was moved to `getSelectedTableCells()` as an initial filter to speed things up everywhere, not only here.

---

To sum up:
* the condition should land at the beginning of `getSelectedTableCells()` (return an empty array if `false`) as a speed benefit for all the sub-features,
* then it's enough to check if `getSelectedTableCells()` has length greater than 1.

Reply:
Sure fair enough. I've forgot about Firefox.

Reply:
I've updated the code.

Reply:
I've updated the code.